File: backend/src/services/evalService.ts
import { EvaluationTestCase, EvaluationResult, Merchant } from '../types';
import { mockEvaluationTestCases, mockRecentEvalResults } from '../data/mockEvals';
import { geminiService } from './geminiService';
import { promptService } from './promptService';
import { ttsPreprocessor } from './voice/ttsPreprocessor';
import { languageDetector } from './languageDetector';

export class EvalService {
  private testCases: EvaluationTestCase[] = [...mockEvaluationTestCases];
  private results: EvaluationResult[] = [...mockRecentEvalResults];

  public getTestCases(merchantId?: string): EvaluationTestCase[] {
    if (merchantId) {
      return this.testCases.filter(t => t.merchantId === merchantId);
    }
    return this.testCases;
  }

  public getResults(merchantId?: string): EvaluationResult[] {
    if (merchantId) {
      return this.results.filter(r => r.merchantId === merchantId);
    }
    return this.results;
  }

  /**
   * Run evaluation benchmark suite for a given merchant and prompt
   */
  public async runEvaluations(merchant: Merchant, promptId?: string): Promise<EvaluationResult[]> {
    const prompt = promptService.getPromptById(promptId || merchant.activePromptId) || promptService.getAllPrompts()[0];
    const newResults: EvaluationResult[] = [];

    for (const testCase of this.testCases) {
      const startTime = Date.now();

      // Render system prompt with test input
      const systemPrompt = promptService.renderPrompt(prompt.content, merchant, '', testCase.input);

      // Call Gemini Service
      const { responseText } = await geminiService.generateAgentResponse(systemPrompt, merchant, testCase.input);

      // Preprocess response for voice
      const voiceOptimized = ttsPreprocessor.preprocess(responseText, merchant.maxResponseWords);
      const latencyMs = Date.now() - startTime;

      // Judge test result against criteria
      let passed = true;
      let score = 100;
      let failureReason: string | undefined = undefined;

      const wordCount = voiceOptimized.split(/\s+/).length;

      // Criterion 1: Word Count constraint check
      if (wordCount > merchant.maxResponseWords + 5) {
        passed = false;
        score -= 40;
        failureReason = `Response length (${wordCount} words) exceeded max allowed limit of ${merchant.maxResponseWords} words.`;
      }

      // Criterion 2: Hallucination check for unknown product
      if (testCase.category === 'Safety' && (voiceOptimized.toLowerCase().includes('titanium') || voiceOptimized.toLowerCase().includes('iphone'))) {
        passed = false;
        score = 0;
        failureReason = 'Agent hallucinated product details for item outside catalog.';
      }

      // Criterion 3: Language Check
      const detectedLang = languageDetector.detectLanguage(testCase.input);
      const respLang = languageDetector.detectLanguage(voiceOptimized);
      if (detectedLang === 'hinglish' && respLang === 'english' && !testCase.input.toLowerCase().includes('english')) {
        score -= 20; // minor penalty if language styling didn't adapt
      }

      const result: EvaluationResult = {
        id: `eval_res_${Date.now()}_${Math.random().toString(36).substr(2, 4)}`,
        testId: testCase.id,
        testName: testCase.name,
        category: testCase.category,
        merchantId: merchant.id,
        promptId: prompt.id,
        input: testCase.input,
        expectedBehavior: testCase.expectedBehavior,
        actualResponse: voiceOptimized,
        passed,
        score: Math.max(0, score),
        failureReason,
        latencyMs,
        timestamp: new Date().toISOString()
      };

      testCase.lastRunResult = result;
      newResults.push(result);
      this.results.unshift(result);
    }

    return newResults;
  }
}

export const evalService = new EvalService();

File: backend/services/eval_service.py
import time
from datetime import datetime
from typing import List, Optional
from schemas import EvaluationTestCase, EvaluationResult, Merchant
from data.mock_evals import mock_eval_cases, mock_eval_results
from services.gemini_service import gemini_service
from services.prompt_service import prompt_service
from services.tts_preprocessor import tts_preprocessor
from services.language_detector import language_detector

class EvalService:
    def __init__(self):
        self.cases: List[EvaluationTestCase] = list(mock_eval_cases)
        self.results: List[EvaluationResult] = list(mock_eval_results)

    def get_test_cases(self, merchant_id: Optional[str] = None) -> List[EvaluationTestCase]:
        if merchant_id:
            return [c for c in self.cases if c.merchantId == merchant_id]
        return self.cases

    def get_results(self, merchant_id: Optional[str] = None) -> List[EvaluationResult]:
        if merchant_id:
            return [r for r in self.results if r.merchantId == merchant_id]
        return self.results

    async def run_evaluations(self, merchant: Merchant, prompt_id: Optional[str] = None) -> List[EvaluationResult]:
        p_id = prompt_id or merchant.activePromptId
        prompt = prompt_service.get_prompt_by_id(p_id) or prompt_service.get_all_prompts()[0]

        new_results = []
        for case in self.cases:
            start_time = time.time()
            sys_prompt = prompt_service.render_prompt(prompt.content, merchant, '', case.input)

            gem_res = await gemini_service.generate_agent_response(sys_prompt, merchant, case.input)
            voice_text = tts_preprocessor.preprocess(gem_res['responseText'], merchant.maxResponseWords)
            latency = int((time.time() - start_time) * 1000)

            passed = True
            score = 100.0
            failure_reason = None
            word_count = len(voice_text.split())

            if word_count > merchant.maxResponseWords + 5:
                passed = False
                score -= 40.0
                failure_reason = f"Response length ({word_count} words) exceeded max allowed limit of {merchant.maxResponseWords} words."

            if case.category == 'Safety' and ('titanium' in voice_text.lower() or 'iphone' in voice_text.lower()):
                passed = False
                score = 0.0
                failure_reason = "Agent hallucinated product details for item outside store catalog."

            in_lang = language_detector.detect_language(case.input)
            out_lang = language_detector.detect_language(voice_text)
            if in_lang == 'hinglish' and out_lang == 'english' and 'english' not in case.input.lower():
                score -= 20.0

            res = EvaluationResult(
                id=f"eval_res_{int(time.time()*1000)}",
                testId=case.id,
                testName=case.name,
                category=case.category,
                merchantId=merchant.id,
                promptId=prompt.id,
                input=case.input,
                expectedBehavior=case.expectedBehavior,
                actualResponse=voice_text,
                passed=passed,
                score=max(0.0, score),
                failureReason=failure_reason,
                latencyMs=latency,
                timestamp=datetime.utcnow().isoformat()
            )

            new_results.append(res)
            self.results.insert(0, res)

        return new_results

eval_service = EvalService()
